File: src/main/java/io/github/faecraft/heartshapedbox/networking/S2CBodyPartSyncPacket.java
package io.github.faecraft.heartshapedbox.networking;

import io.github.faecraft.heartshapedbox.body.AbstractBodyPart;
import io.github.faecraft.heartshapedbox.body.BodyPartProvider;
import net.fabricmc.fabric.api.networking.v1.PacketByteBufs;
import net.fabricmc.fabric.api.networking.v1.ServerPlayNetworking;
import net.minecraft.network.PacketByteBuf;
import net.minecraft.server.network.ServerPlayerEntity;
import net.minecraft.util.Identifier;

import java.util.HashSet;
import java.util.Optional;

import static io.github.faecraft.heartshapedbox.main.HSBMain.MOD_ID;

public class S2CBodyPartSyncPacket {
    public static final Identifier IDENTIFIER = new Identifier(MOD_ID, "body_part_sync");
    private final HashSet<AbstractBodyPart> parts = new HashSet<>();

    public S2CBodyPartSyncPacket() {}

    public S2CBodyPartSyncPacket(BodyPartProvider provider) {
        parts.addAll(provider.getParts());
    }

    public void addPart(AbstractBodyPart part) {
        parts.add(part);
    }

    public static void update(PacketByteBuf buffer, BodyPartProvider provider) {
        while (buffer.isReadable()) {
            Identifier id = buffer.readIdentifier();
            float health = buffer.readFloat();
            float maxHealth = buffer.readFloat();

            Optional<AbstractBodyPart> optionalPart = provider.maybeGet(id);

            if (optionalPart.isPresent()) {
                AbstractBodyPart part = optionalPart.get();

                part.setHealth(health);
                part.setMaxHealth(maxHealth);
            }
        }
    }

    public static S2CBodyPartSyncPacket from(ServerPlayerEntity player) {
        return new S2CBodyPartSyncPacket((BodyPartProvider) player);
    }

    public void write(PacketByteBuf buffer) {
        for (AbstractBodyPart part : parts) {
            buffer.writeIdentifier(part.getIdentifier());
            buffer.writeFloat(part.getHealth());
            buffer.writeFloat(part.getMaxHealth());
        }
    }

    public void send(ServerPlayerEntity player) {
        PacketByteBuf buffer = PacketByteBufs.create();

        write(buffer);

        ServerPlayNetworking.send(player, IDENTIFIER, buffer);
    }
}

File: src/main/kotlin/io/github/faecraft/heartshapedbox/networking/S2CBodyPartSyncPacket.kt
package io.github.faecraft.heartshapedbox.networking

import io.github.faecraft.heartshapedbox.body.AbstractBodyPart
import io.github.faecraft.heartshapedbox.body.BodyPartProvider
import io.github.faecraft.heartshapedbox.main.HSBMain
import net.fabricmc.fabric.api.networking.v1.PacketByteBufs
import net.fabricmc.fabric.api.networking.v1.ServerPlayNetworking
import net.minecraft.network.PacketByteBuf
import net.minecraft.server.network.ServerPlayerEntity
import net.minecraft.util.Identifier
import org.apache.logging.log4j.LogManager
import java.util.*

public class S2CBodyPartSyncPacket() {
    private val parts = HashSet<AbstractBodyPart>()

    public constructor(provider: BodyPartProvider) : this() {
        parts.addAll(provider.parts)
    }

    public fun addPart(part: AbstractBodyPart) {
        parts.add(part)
    }

    private fun write(buffer: PacketByteBuf) {
        for (part in parts) {
            buffer.writeIdentifier(part.getIdentifier())
            buffer.writeFloat(part.getHealth())
            buffer.writeFloat(part.getMaxHealth())
        }
    }

    public fun send(player: ServerPlayerEntity) {
        // TODO: Remove logging call
        LOGGER.info("Sending packet to player: " + player.displayName.asString())
        val buffer = PacketByteBufs.create()
        write(buffer)
        ServerPlayNetworking.send(player, IDENTIFIER, buffer)
    }

    public companion object {
        private val LOGGER = LogManager.getLogger("S2CBodyPartSyncPacket")
        public val IDENTIFIER: Identifier = Identifier(HSBMain.MOD_ID, "body_part_sync")

        public fun update(buffer: PacketByteBuf, provider: BodyPartProvider) {
            while (buffer.isReadable) {
                val id = buffer.readIdentifier()
                val health = buffer.readFloat()
                val maxHealth = buffer.readFloat()
                // TODO: Remove logging calls
                LOGGER.info("UPDATE | Provider: $provider")
                LOGGER.info("UPDATE | ID: $id Health: $health Max Health: $maxHealth")
                val optionalPart = provider.maybeGet(id)
                if (optionalPart.isPresent) {
                    val part = optionalPart.get()
                    part.setHealth(health)
                    part.setMaxHealth(maxHealth)
                }
            }
        }

        @JvmStatic
        public fun from(player: ServerPlayerEntity): S2CBodyPartSyncPacket =
            S2CBodyPartSyncPacket(player as BodyPartProvider)
    }
}
